File: backend/app/schemas/substance/substance.py
import decimal
from typing import Optional

from pydantic import BaseModel, ConfigDict

from app.schemas.department import DepartmentRead
from app.schemas.substance.hazard_category import HazardCategoryBase
from app.schemas.substance.property import PropertyBase


class SubstanceBase(BaseModel):
    name: str
    mixture: bool = True
    physical_form_name: Optional[str] = None
    unit_name: Optional[str] = None
    sds_revision_year: Optional[int] = None
    note: Optional[str] = None
    water_toxicity_ec50: Optional[str] = None
    manufacturer: Optional[str] = None
    code: Optional[str] = None
    company_id: Optional[int] = None
    sds: Optional[str] = None
    hazard_category: list[HazardCategoryBase] = []

class SubstanceCreate(SubstanceBase):
    property_ids: list[int] = []

class SubstanceUpdate(SubstanceBase):
    property_ids: list[int] = []

class SubstanceRead(SubstanceBase):
    id: int
    properties: Optional[list[PropertyBase]] = None
    departments: list[SubstanceDepartments] = []

    model_config = ConfigDict(from_attributes=True)


class SubstancePaginationRead(BaseModel):
    items: list[SubstanceRead]
    total: int
    limit: int
    offset: int


class SubstanceDepartments(BaseModel):
    department: DepartmentRead
    year: int
    amount: decimal.Decimal

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/schemas/Substance.ts
import type {Property} from "./Property.js";
import type {SubstanceDepartment} from "./SubstanceDepartment.ts";
import type {HazardCategory} from "./HazardCategory.js";

export interface SubstanceRead {
    id: number;
    name: string;
    mixture: boolean;
    physical_form_name?: string;
    unit_name?: string;
    properties?: Property[];
    sds_revision_year?: number;
    note?: string;
    water_toxicity_ec50?: string;
    manufacturer?: string;
    code?: string;
    company_id?: number;
    sds?: string;
    departments: SubstanceDepartment[];
    hazard_category: HazardCategory[];
}

export interface SubstanceCreate {
    name: string;
    mixture: boolean;
    physical_form_name?: string;
    unit_name?: string;
    property_ids?: number[];
    sds_revision_year?: number;
    note?: string;
    water_toxicity_ec50?: string;
    manufacturer?: string;
    code?: string;
    company_id?: number;
    sds?: string;
    hazard_category: HazardCategory[];
}

export interface SubstancePaginationRead {
    items: SubstanceRead[];
    total: number;
    limit: number;
    offset: number;
}